New module: Custom HUD Components

fracticality opened the pull request:
Allows a modder to easily inject a custom class as a HUD component

fracticality committed:
Update vmf_loader.lua

fracticality committed:
Add files via upload

fracticality committed:
Update vmf_loader.lua

snezhniylis commented:
![image](https://user-images.githubusercontent.com/85341530/173088506-b6959180-725d-45db-9702-7c3d4a7f1b54.png)

That's an awesome contribution, thank you!

I see there's a VT1 check at the bottom, so, I assume, it works with both games? I suppose, the functionality has already been battle-tested, so I can safely merge it. Currently, I'm unable to test this myself, so I'll trust you on this one. Will merge and apply for sanctioning tomorrow.

Dreomite commented:
![image](https://user-images.githubusercontent.com/35058624/173092159-1a7395ec-e083-4ef5-ae7b-82b1ee3909f7.png)

This is an amazing contribution, thank you!

I see there's a VT1 check at the bottom, so I assume this works with both games? I suppose, the functionality has already been battle-tested, so I can safely merge it. Since I'm unable to test it right now, I'll trust you on this one. Will merge and apply for sanctioning tomorrow.

Dreomite commented:
I decided to do a proper code review before merging this, and 2 things caught my attention:

1. Vermintide 1 HUD have no such thing as components. This API is VT2-exclusive.
2. It neither injects any custom resources, not grants any benefits over doing this non-VMF way.

This change makes modding Vermintide 2 HUD much more convenient, that's for sure. But what bothers me, is that it looks to be a bit outside of the VMF scope. There are somewhat similar modules in VMF, but they all have strong reasons to be used inside the framework:

- `custom_textures` - Used internally for Mod Options view; extremely convoluted ui_renderer injecting process makes it rather necessary to be handled on the VMF side for compatibility purposes
- `custom_views` - Used internally for Mod Options view; has keybindings integration for toggling; compatibility on top of it is just a bonus

So I'm really torn apart right now. It's hard to hide, that I'm very excited about this recent VMF activity, and it would also be hard to refuse the obviously convenient changes you've put so much effort in. But, on the other hand, this adds to the burden of maintenance, for something that can be easily done outside the framework. That's why, for example, grasmann's Simple UI was split into a different mod, or Aussiemon's High DPI fix. (And, I'ma be honest, that's partially what made me hesitant of merging #37)

Have this been discussed with the modding community in the discord server? I've been taking a long break from all my social media, so I haven't visited the server for quite some time.

I don't feel like I can make this decision alone. I'd like to ask other the core contributors what they think about merging this: @ManuelBlanc @sclu1034 @FireSiku @Aussiemon (obviously, feel free to ignore)

FireSiku commented:
Coming back from the dead to give my two cents on the topics of the VMF scope and it's inclusion:

I've always been a strong proponent for VMF to have a series of standardized ways of doing things that makes it more convenient for mod authors to bring their ideas into reality. While people can do this in a "non-VMF" way, if all of them are required to hook the same functions, individually keep track of the ingame_hud object and so on, then it absolutely make sense for VMF as a framework to provide a standardized method for everyone to use.

This reduces the risk of certain mods negatively impacting others (ie: what if one mod's implementation uses hook instead of hook_safe?) and as a community, it makes it easier to help mod authors troubleshoot Custom HUD errors if we dont have to question their implementation on top of it. 

In regards to maintainability, the code is concise (less than 200 lines), mostly pure lua with some unavoidable exceptions to reduce the likelihood of the code breaking with a new VT2 patch, and uses clear variable names. That makes it easy to maintain because any contributor could pick up the mantle should fracticality be absent for a long period of time. The code doesn't look to have an high overhead for players who arent using any mod that makes uses of these functions, which is good.

The only thing I would recommend might be some additional block comments on some of it works. For example, we have a function to validate `component_settings`, but it's not immediately clear to me what this component_settings should contain, and which parts of it are maintained or specified by VMF and which parts, if any, are actually required to be strictly adhered to as required by the VT client itself. (Ex: "If visibility_group is missing, VT client will crash")

Last topic I will touch is about its inclusion when compared to other things that were originally moved out of VMF:
Aussiemon's High DPI Fix: This one is quite easy, because it pertains to the user experience directly. Having that code in means the end user's experience is altered. It doesn't improve the quality of life for mod authors to make new mods. It's not used as a dependency as far as I'm aware, even if it is highly recommended to use by some. Makes sense for it to be a standalone mod for users to download and not be a performance burden enforced to everyone using VMF.

SimpleUI: This one is a bit trickier because it does look to be focused on being a library to help mod authors do something and its possible the design philosophy when we decision was made were different. Looking at the code for it, the sheer size of it (being 4000+ lines) felt daunting to me and it looked like several hundred tables would be generated and populated, regardless of a end user having any needs for it which is overhead that should be avoided in a framework that is mandatory by nearly every mods.

fracticality committed:
custom_hud_components.lua
- Cleaned up error message table.
- Added documentation regarding expected component settings.
- Removed VT1 check for ingame_hud, as components are not compatible with its HUD.
- Reduced line lengths to <120 characters to nuke build warnings

vmf_loader.lua
- custom_hud_components is now only (re)loaded in VT2.

fracticality committed:
custom_hud_components.lua
- Cleaned up error message table.
- Added documentation regarding expected component settings.
- Removed VT1 check for ingame_hud, as components are not compatible with its HUD.
- Reduced line lengths to <120 characters to nuke build warnings

vmf_loader.lua
- custom_hud_components is now only (re)loaded in VT2.

sclu1034 commented:
I agree with @FireSiku that making things convenient is part of the job of a project like VMF. Conceptually, I'm fine with adding this.

sclu1034 reviewed (changes requested):
One thing I would like to see added, though, is for VMF to add/remove components in the HUD (or at least hide it) automatically when the mod is enabled/disabled, respectively.

Dreomite commented:
@FireSiku
What you wrote, totally makes sense and I agree with it. Thank you so much for your concise and well-thought-out answer!

@sclu1034
Thanks for voicing your opinion!

@fracticality
@sclu1034's proposed functionality sounds reasonable. If it's relatevely pain-free and won't impede modders in any way, it would be great if you implemented it. Other than that, this PR is ready for merging.

Optionally, you could change you module's indentation to 2 spaces to match the rest of the project, but that's a non-blocker.

fracticality commented:
I'd be glad to implement automatic injection/removal based on mod enabled state.

I also would like to hear opinions on whether implementing custom visibility groups would be too much. Having such an implementation would make it easy to, for example, show a component based purely on another component's visibility rather than relying on finding all the right groups (which are now spread out across many files due to DLC).
Creating such a group is simply providing a name and a validation function.

sclu1034 commented:
> implementing custom visibility groups

Good idea.
I don't think your specific example is what those are intended for (scenegraphs should be used to create such dependencies), but the actual contents of a validation function would be up to the individual mod anyways.

We would have to make sure to handle name duplicates across mods, though.

fracticality commented:
Custom visibility groups sounded great in theory, but in practice doesn't seem good:
- Only one visibility group can be active at a time
- The active visibility group is chosen on a first-come, first-serve basis
If a custom visibility group is placed after all the others, it runs a high risk of never being active.
If a custom visibility group is placed before the others (or somewhere in the middle), it runs a high risk of preventing a needed group from becoming active.

Thus, I feel it'd be better to forgo that functionality.

fracticality committed:
toggling.lua
- Added function calls to inject/remove hud components on mod toggle

custom_hud_components.lua
- Added 'status' enum to better enforce injection state
- Two-space tabs for style consistency
- Added methods of injecting/removing components for a specific mod

fracticality commented:
Assuming everyone is comfortable with the latest commits, I'm comfortable saying it's done. :)

Dreomite commented on vmf/scripts/mods/vmf/modules/core/toggling.lua:
```diff
@@ -13,9 +13,11 @@ vmf.set_mod_state = function (mod, is_enabled, initial_call)
   if is_enabled then
     mod:enable_all_hooks()
     vmf.mod_enabled_event(mod, initial_call)
+    vmf.inject_hud_components(mod)
```
The injecting should be done before `on_enabled` event is called. So, when it's called, the modder is able to access and do whatever with already injected components.

fracticality replied:
Sure thing!

Dreomite commented on vmf/scripts/mods/vmf/modules/core/toggling.lua:
```diff
@@ -13,9 +13,11 @@ vmf.set_mod_state = function (mod, is_enabled, initial_call)
   else
     mod:disable_all_hooks()
     vmf.mod_disabled_event(mod, initial_call)
+    vmf.remove_injected_hud_components(mod)
```
Would it be safe to move this up as well? The idea is to call `on_disabled` after all the internal changes inside the mod have been done, so the modder has complete power over the final state of the mod when it's disabled.

fracticality replied:
Safe for this as well.

Dreomite reviewed (changes requested):
Looking good. Just 2 things that need addressing. I haven't delved into the HUD code myself, so maybe it makes sense to do things the way you did. Other than that, this PR is ready for merging.

fracticality committed:
custom_hud_components.lua.lua
- Component status is reset on `IngameHud.destroy` to ensure components are reinjected properly when party members change.

toggling.lua
- Inject/remove events now executed before respective enabled/disabled event hooks.

fracticality committed:
Revert "custom_hud_components.lua.lua - Component status is reset on `IngameHud.destroy` to ensure components are reinjected properly when party members change."

This reverts commit 2384a43b

fracticality commented:
Implemented your requests (and fixed a rather embarrassing oversight).

fracticality committed:
custom_hud_components.lua
- Component status is reset on `IngameHud.destroy` to ensure components are reinjected properly when party members change.

toggling.lua
- Inject/remove events now executed before respective enabled/disabled event hooks.

Dreomite commented:
Thank you for your amazing work, Fracticality! I've already updated VMF Beta in Steam Workshop. I'm going to squash-merge this PR and apply for sanctioning.

Dreomite commented:
@fracticality I've uploaded the sanctioned version to workshop (also, added you as a contributor there).

It would be nice if someone did a little PSA in Vermintide Modders discord about this new functionality in any channel they see fits best.

fracticality commented:
Great news!
Once I've updated the docs I'll handle a PSA.

Thank you!
